django_auth_ldap/backend.py

```diff
@@ -45,94 +45,146 @@
-from django.contrib.auth.models import Group, Permission
+from django.contrib.auth.models import AbstractUser, Group, Permission
 from django.core.cache import cache
 from django.core.exceptions import ImproperlyConfigured, ObjectDoesNotExist
+from django.http import HttpRequest
 from django.utils.inspect import func_supports_parameter
+from ldap.ldapobject import LDAPObject
 
 from django_auth_ldap.config import (
+    AbstractLDAPSearch,
     ConfigurationWarning,
     LDAPGroupQuery,
+    LDAPGroupType,
     LDAPSearch,
     _LDAPConfig,
 )
 
-logger = _LDAPConfig.get_logger()
+try:
+    from django.contrib.auth.backends import BaseBackend
+except ImportError:
+    BaseBackend = object
 
+T: TypeVar = TypeVar("T")
+
+logger: Logger = _LDAPConfig.get_logger()
 
 # Exported signals
 
 # Allows clients to perform custom user population.
 # Passed arguments: user, ldap_user
-populate_user = django.dispatch.Signal()
+populate_user: django.dispatch.Signal = django.dispatch.Signal()
 
 # Allows clients to inspect and perform special handling of LDAPError
 # exceptions. Exceptions raised by handlers will be propagated out.
 # Passed arguments: context, user, exception
-ldap_error = django.dispatch.Signal()
+ldap_error: django.dispatch.Signal = django.dispatch.Signal()
 
 
-class LDAPBackend:
+class LDAPBackend(BaseBackend):
     """
     The main backend class. This implements the auth backend API, although it
     actually delegates most of its work to _LDAPUser, which is defined next.
     """
 
-    supports_anonymous_user = False
-    supports_object_permissions = True
-    supports_inactive_user = False
-
-    _settings = None
-    _ldap = None  # The cached ldap module (or mock object)
-
-    # This is prepended to our internal setting names to produce the names we
-    # expect in Django's settings file. Subclasses can change this in order to
-    # support multiple collections of settings.
-    settings_prefix = "AUTH_LDAP_"
-
-    # Default settings to override the built-in defaults.
-    default_settings = {}
-
-    def __getstate__(self):
+    def __init__(
+        self,
+        supports_anonymous_user: bool = False,
+        supports_object_permissions: bool = True,
+        supports_inactive_user: bool = False,
+        settings: Optional[LDAPSettings] = None,
+        settings_prefix: str = "AUTH_LDAP_",
+        default_settings: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        # Check hasattr for support of previous class variables.
+
+        if not hasattr(self, "_settings"):
+            self.supports_anonymous_user: bool = supports_anonymous_user
+        if not hasattr(self, "_settings"):
+            self.supports_object_permissions: bool = supports_object_permissions
+        if not hasattr(self, "_settings"):
+            self.supports_inactive_user: bool = supports_inactive_user
+
+        if not hasattr(self, "_settings"):
+            self._settings: Optional[LDAPSettings] = settings
```

Comment:
These `if` can be combined, no need to repeat the check.

Reply:
Thanks for the hint. No they should not be combined neither should they all check `hasattr(self, "_settings")`, instead they should check for their corresponding attribute.